zerver/actions/message_send.py

```diff
@@ -857,6 +863,101 @@ def get_active_presence_idle_user_ids(
     return filter_presence_idle_user_ids(user_ids)
 
 
+def insert_idempotent_msg(send_request: SendMessageRequest) -> None:
+    """
+    Insert a single Message row in db, ensuring Idempotency using Idempotency-Key header.
+    """
+
+    field_names = [
+        "sender_id",
+        "recipient_id",
+        "realm_id",
+        "type",
+        "subject",
+        "content",
+        "rendered_content",
+        "rendered_content_version",
+        "date_sent",
+        "sending_client_id",
+        "last_edit_time",
+        "edit_history",
+        "has_attachment",
+        "has_image",
+        "has_link",
+        "is_channel_message",
+        "search_tsvector",
+    ]
```

Comment:
It'd be good to figure out how to not have to hardcode this list. If you forgot a field here, does the test fail? 

I bet Django internals offer a way to ask for the SQL you need.

Reply:
Yes, it's possible to generate those fields dynamically

Reply:
I think there's bigger issue here. How can we verify that creating a model object with this method produces the same `Message` row in Postgres as doing so with the `bulk_create` code path?

Django's real `bulk_create` has a bunch of logic in functions like  `_prepare_related_fields_for_save`.

Reply:
> Yeah. For instance, I think this will not apply `default` values, which are applied in Django, not the database?

No. The default values are already applied to the `Message` instnace, and I get those values in:

```python
column_to_value: dict[str, Any] = {
    col_name: getattr(message, col_name) for col_name in field_names
}
```